### PR TITLE
Flexible augmentation config for Lhotse dataloaders

### PR DESCRIPTION
**Pull request by pzelasko**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds two new configuration fields `augment_examples` and `augment_batch` that let the user specify their own augmentations in a custom order.

This is following a design descussion with @stevehuang52. 

**Collection**: ASR, Audio, speechlm2

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by stevehuang52** [comment]
Is it possible to add an example on how to wrap an augmentation function performed on torch.tensor or np.ndarray into something compatible with lhotse?

**Comment by pzelasko** [comment]
> Is it possible to add an example on how to wrap an augmentation function performed on torch.tensor or np.ndarray into something compatible with lhotse?

Yeah. I always point people to this PR in Lhotse as a reference for how to fully add a new augmentation:
https://github.com/lhotse-speech/lhotse/pull/382/files#diff-add451896faa625c1820580ab6ad64bef75e2886d551efc0f5705100ea62b28a

When adding a new Lhotse-compatible augmentation in NeMo, you need only two steps from that PR:
* `torchaudio.py`: similarly to `Volume`, inherit your transform from `AudioTransform` and manipulate np.ndarray there
* `audio.py`: define a function similar to `Recording.perturb_volume` which adds desired transform to `Recording.transforms` (in Lhotse it returns a copy of `Recording`, but if you feel it's adequate, feel free to modify the objects in-place). It can be a free function (as opposed to a method).

The changes in `cut.py` are just to be able to maintain immutability of the original objects by creating shallow copies of recordings, supervisions, cuts, etc. This is a requirement in Lhotse, but permissible to skip in user code.

File: nemo/collections/common/data/lhotse/augment.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+                aug = item
+            else:
+                assert isinstance(item, (dict, DictConfig))
+                aug = hydra.utils.instantiate(item)
+            sampler = sampler.map(aug)
+        return sampler
+    return augment_batch_legacy(sampler, config)
+
+
+def augment_examples_legacy(cuts: CutSet, config: DictConfig) -> CutSet:
```
Comment (anteju) [inline nemo/collections/common/data/lhotse/augment.py:68]: Could we add a docstring with `config.augment_examples` which would be equivalent to `augment_examples_legacy`?
That could serve as a basic example of using the new config from this PR.

File: nemo/collections/common/data/lhotse/augment.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+        if callable(item):
+            aug = item
+        else:
+            assert isinstance(item, (dict, DictConfig))
+            aug = hydra.utils.instantiate(item)
+        cuts = aug(cuts)
+    return cuts
+
+
+def augment_batch(sampler: CutSampler, config: DictConfig) -> CutSampler:
```
Comment (anteju) [inline nemo/collections/common/data/lhotse/augment.py:54]: Add docstring with an example.

File: nemo/collections/common/data/lhotse/augment.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            keep_excessive_supervisions=config.keep_excessive_supervisions,
+        )
+
+    if config.pad_min_duration is not None:
+        cuts = cuts.pad(duration=config.pad_min_duration, direction=config.pad_direction, preserve_id=True)
+
+    return cuts
+
+
+def augment_batch_legacy(sampler: CutSampler, config: DictConfig) -> CutSampler:
```
Comment (anteju) [inline nemo/collections/common/data/lhotse/augment.py:111]: Similar as for `augment_examples_legacy`, consider adding `config.augment_batch` which would be equivalent to legacy behavior.

File: nemo/collections/common/data/lhotse/augment.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        return augment_examples_legacy(cuts, config)
+    assert isinstance(config.augment_examples, (list, ListConfig))
+    for item in config.augment_examples:
+        if callable(item):
+            aug = item
+        else:
+            assert isinstance(item, (dict, DictConfig))
+            aug = hydra.utils.instantiate(item)
+        cuts = aug(cuts)
+    return cuts
```
Comment (anteju) [inline nemo/collections/common/data/lhotse/augment.py:51]: Consider adding debug logs, so it's easier to debug any issues by changing the logging level.
Similar for `augment_batch` and other functions.

**Comment by stevehuang52** [comment]
> * `audio.py`: define a function similar to `Recording.perturb_volume` which adds desired transform to `Recording.transforms` (in Lhotse it returns a copy of `Recording`, but if you feel it's adequate, feel free to modify the objects in-place). It can be a free function (as opposed to a method).

It seems that adding something like `Recording.perturb_volume` needs to change lhotse itself, can we do it in NeMo without changing lhotse?

**Comment by stevehuang52** [comment]
Could you please also add relevant documentation to https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/datasets.html#lhotse-dataloading?

File: nemo/collections/common/data/lhotse/augment.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+    if not config.augment_examples:
+        return augment_examples_legacy(cuts, config)
+    assert isinstance(config.augment_examples, (list, ListConfig))
+    for item in config.augment_examples:
+        if callable(item):
+            aug = item
+        else:
+            assert isinstance(item, (dict, DictConfig))
+            aug = hydra.utils.instantiate(item)
+        cuts = aug(cuts)
```
Comment (stevehuang52) [inline nemo/collections/common/data/lhotse/augment.py:50]: Does the `aug` function perform augmentation by looping over all `Cut` objects in `CutSet`? If so, would this slow down datlaloading? If this function is to apply augmentation on each individual sample, it seems that the input should be a `Cut` instead of `CutSet`, then `augment_batch` can reuse `augment_sample` to apply it on a batch of data?

**Comment by stevehuang52** [comment]
Another question not directly related to this PR: I think the current main limitation of Lhotse is that we still sometimes have to do some operations in the dataset's `__getitem__` function that require a for-loop implementation, while customized augmentation is just one of them. Any suggestions on solving the for-loop issue in general? Writing everything as "augmentation" may be a workaround but doesn't seem to be a neat solution.

File: nemo/collections/common/data/lhotse/augment.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        )
+
+    # 2.b. On-the-fly speed perturbation.
+    #    mux here ensures it's uniformly distributed throughout sampling,
+    #    and applying it here (before sampler/dataset) ensures optimal
+    #    bucket allocation.
+    if config.perturb_speed:
+        cuts = CutSet.mux(
+            cuts,
+            cuts.perturb_speed(0.9),
```
Comment (stevehuang52) [inline nemo/collections/common/data/lhotse/augment.py:87]: Does this mean either apply `perturb_speed(0.9)` and `perturb_speed(1.1)` to all samples with `prob=1.0` and therefore making the dataset 3x as large as before? How to do it randomly for each sample, such that each time every sample has a certain probability to be augmented?

File: nemo/collections/common/data/lhotse/augment.py (hunk 5 above)
Comment (stevehuang52) [reply nemo/collections/common/data/lhotse/augment.py:50]: What's the difference between `cuts = aug(cuts)` and `cuts.map(aug)`, and how to decide which to use?

**Comment by nithinraok** [comment]
@pzelasko is this PR still relevant
